File: reconst_admm.py
import os
import numpy as np
import cupy as cp
import scipy.sparse as ssp
from PIL import Image
from package import myUtil
import admm

DATA_PATH = "../data"

# OBJ_NAMES = ["White", "Cameraman", "Text", "Daruma", "Woman"]
OBJ_NAMES = ["Woman"]

H_SETTINGS = ["gf", "p-5_lmd-100_to-False", "int_p-5_lmd-100_to-True"]

CAP_DATE = "241114"
EXP_DATE = "241127"
n = 128
m = 255


def create_D_mono(n):
    I = ssp.eye(n**2, format="csr")

    Dx = I - ssp.csr_matrix(np.roll(I.toarray(), 1, axis=1))
    Dx[n - 1 :: n, :] = 0
    Dy = I - ssp.csr_matrix(np.roll(I.toarray(), n, axis=1))
    Dy[-n:, :] = 0

    return ssp.vstack([Dx, Dy])


D = create_D_mono(n)
D = cp.array(D.toarray())

# すべてのOBJ_NAMEとH_SETTINGの組み合わせを処理
for OBJ_NAME in OBJ_NAMES:
    captured_path = f"{DATA_PATH}/capture_{CAP_DATE}/{OBJ_NAME}.png"

    captured = Image.open(captured_path).convert("L")
    captured = cp.asarray(captured)
    black = myUtil.calculate_bias(m**2, DATA_PATH, CAP_DATE)
    g = captured.ravel() - black

    for H_SETTING in H_SETTINGS:
        H_matrix_path = f"{DATA_PATH}/{EXP_DATE}/systemMatrix/H_matrix_{H_SETTING}.npy"

        H = cp.load(H_matrix_path).astype(cp.float32)
        print(f"Processing OBJ_NAME: {OBJ_NAME}, H_SETTING: {H_SETTING}")
        print("H shape:", H.shape, "type(H):", type(H), "H.dtype:", H.dtype)

        admm_solver = admm.Admm(H, g, D)
        f, err = admm_solver.solve()

        f = cp.clip(f, 0, 1)
        f = cp.asnumpy(f.reshape(n, n))
        f_image = Image.fromarray((f * 255).astype(np.uint8), mode="L")

        tau = np.log10(admm_solver.tau)
        mu1 = np.log10(admm_solver.mu1)
        mu2 = np.log10(admm_solver.mu2)
        mu3 = np.log10(admm_solver.mu3)

        reconst_dir = f"{DATA_PATH}/{EXP_DATE}/reconst"
        if not os.path.exists(reconst_dir):
            os.makedirs(reconst_dir)

        SAVE_PATH = f"{reconst_dir}/{OBJ_NAME}_{H_SETTING}_admm_t-{tau:.2f}_m{mu1:.2f}m{mu2:.2f}m{mu3:.2f}.png"
        f_image.save(SAVE_PATH, format="PNG")
        print(f"Saved: {SAVE_PATH}")

File: reconst_admm.m
%% パラメータの設定
DATA_PATH = '../../OneDrive - m.titech.ac.jp/Lab/data';
% DATA_PATH = '../data';
OBJ_NAME = 'Cameraman';
H_SETTING = 'p-5_lmd-100_m-128';
% H_SETTING = 'gf';
CAP_DATE = '241114';
EXP_DATE = '241118';
n = 128;
m = 128;
PREFIX = '';
% PREFIX = 'int_';

%% Dの作成
D = create_D_mono(n);

%% 画像の読み込みと前処理
image_path = fullfile(DATA_PATH, ['capture_' CAP_DATE], [OBJ_NAME '.png']);
captured = imread(image_path);
% resize
captured = imresize(captured, [m, m]);

if size(captured, 3) == 3
    captured = rgb2gray(captured);
end

captured = cast(captured, 'double') / 255;
g = captured(:);

%% システム行列Hの読み込み
H_path = fullfile(DATA_PATH, EXP_DATE, 'systemMatrix', ['H_matrix_' PREFIX H_SETTING '.mat']);
data = load(H_path);
H = cast(data.H, 'double');
% H が疎行列であることの確認
if issparse(H)
    fprintf('H は疎行列です。サイズ: %dx%d\n', size(H, 1), size(H, 2));
else
    fprintf('H は疎行列ではありません。疎行列に変換します。\n');
    H = sparse(H);
end

% Hの情報を表示
fprintf('H shape: [%d, %d]\n', size(H, 1), size(H, 2));
fprintf('Class of H: %s\n', class(H));
fprintf('Data type of H: %s\n', class(H));

%% 追加の処理（必要に応じて）
disp('システム行列H_sparseの概要:');
whos H

%% ADMMソルバーの実行
admm_solver = Admm(H, g, D);
[f_solution, error_history] = admm_solver.solve();

%% 結果の保存
f_solution = min(max(f_solution, 0), 1);
f_solution = reshape(f_solution, [n, n]);
f_image = cast(f_solution * 255, 'uint8');

tau_log = log10(admm_solver.tau);
mu1_log = log10(admm_solver.mu1);
mu2_log = log10(admm_solver.mu2);
mu3_log = log10(admm_solver.mu3);

reconst_dir = fullfile(DATA_PATH, EXP_DATE, 'reconst');

if ~exist(reconst_dir, 'dir')
    mkdir(reconst_dir);
end

save_filename = sprintf('%s_%s_admm_t-%.1f_m%.1f%.1f%.1f.png', ...
    OBJ_NAME, H_SETTING, tau_log, mu1_log, mu2_log, mu3_log);
save_path = fullfile(reconst_dir, save_filename);
imwrite(f_image, save_path, 'png');
fprintf('Image saved to: %s\n', save_path);

function D = create_D_mono(n)
    % スパースの単位行列を作成
    I = speye(n ^ 2);

    %% Dxの作成
    % 行方向に1シフト（列を右に1シフト）
    shifted_Ix = circshift(I, [0, 1]);

    % 最後の列でのラップを防ぐために特定の行をゼロにする
    rows_to_zero_x = n:n:n ^ 2;
    shifted_Ix(rows_to_zero_x, :) = 0;

    % Dxを計算
    Dx = I - shifted_Ix;

    %% Dyの作成
    % 列方向にnシフト（行を下に1シフト）
    shifted_Iy = circshift(I, [0, n]);

    % 最後のn行でのラップを防ぐためにこれらの行をゼロにする
    shifted_Iy(end - n + 1:end, :) = 0;

    % Dyを計算
    Dy = I - shifted_Iy;

    %% DxとDyを垂直に結合してDを作成
    D = [Dx; Dy];
end
